kore/src/Kore/Syntax/Application.hs

```diff
@@ -79,20 +71,18 @@ data Application head child = Application
     { applicationSymbolOrAlias :: !head
     , applicationChildren      :: [child]
     }
-    deriving (Eq, Functor, Foldable, Ord, Traversable, GHC.Generic, Show)
-
-instance (Hashable head, Hashable child) => Hashable (Application head child)
-
-instance (NFData head, NFData child) => NFData (Application head child)
-
-instance SOP.Generic (Application head child)
-
-instance SOP.HasDatatypeInfo (Application head child)
+    deriving (Eq, Ord, Show)
+    deriving (Functor, Foldable, Traversable)
+    deriving (GHC.Generic)
+    deriving anyclass (Hashable, NFData)
+    deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
 
 instance (Debug head, Debug child) => Debug (Application head child)
 
 instance
-    ( Debug head, Debug child, Diff head, Diff child )
+    ( Debug head, Diff head
+    , Debug child, Diff child
+    )
```

Comment:
Same question as above.

Reply:
The compiler can't infer this constraint, but it's for a different reason: There is an overloaded instance of `Debug` for `String` (so that the instance for `String` is different from the instance for `[a]`). There is a `[child]` in `Application`, so instance resolution is ambiguous. If `child ~ Char` then the compiler could pick the `String` instance, but not in the general case. By adding the `Debug child` constraint, we force the compiler to pick the general instance.